### PR TITLE
ACE2ERA5 Perf Improvments

### PR DESCRIPTION
**Pull request by NickGeneva**
# Earth2Studio Pull Request

## Description

Improves ACE2ERA5 forecast performance by caching forcing data in the model wrapper and fixes ACE2ERA5 forcing data retrieval for static variables requested across multiple times.

Key changes:
- Cache ACE2ERA5 forcing data by year on the requested device, replacing the previous per-step forcing fetch path for the default ACE2 datasource.
- Keep custom forcing data sources on the smaller per-time cache path.
- Fix ACE2ERA5 static forcing fields so mixed static, time-only, and dynamic variables can be fetched for multiple timestamps in one call.
- Update ACE2 API docs to remove the paper-derived throughput claim and note `torch.set_float32_matmul_precision("high")` as a useful performance setting.
- Add regression coverage for multi-time static forcing retrieval.

<details>
<summary>Performance notes and benchmark script</summary>

Measured on the Colossus VM with CUDA using the local `test.py` script below. The loop timing excludes model/package loading and reports the ACE2 iterator loop for `loop[28]`.

| Configuration | Iterator loop time | Relative to baseline |
| --- | ---: | ---: |
| Baseline before forcing cache | ~2.52 s | 1.00x |
| Per-timestep forcing cache + TF32 + `torch.compile` | ~1.45 s | ~1.74x faster |
| Yearly forcing cache + TF32 + `torch.compile` | ~1.38 s | ~1.83x faster |

The current branch measured `Elapsed time(loop[28]): 1.380558 seconds`. The one-time `torch.compile` warmup was about `10.6 seconds` and is reported separately by the script.

```python
import time

import numpy as np
import torch
from tqdm import tqdm

from earth2studio.data import fetch_data
from earth2studio.data.ace2 import ACE2ERA5Data
from earth2studio.models.px import ACE2ERA5
from earth2studio.utils.time import to_time_array

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
torch.set_float32_matmul_precision("high")
torch.backends.cudnn.benchmark = True
use_torch_compile = True

start_date = "1950-01-01"


data = ACE2ERA5Data()


package = ACE2ERA5.load_default_package()

prognostic = ACE2ERA5.load_model(package).to(device)

if use_torch_compile:
    prognostic.stepper.modules[0].module = torch.compile(
        prognostic.stepper.modules[0].module, mode="reduce-overhead"
    )


x, coords = fetch_data(
    source=data,
    time=to_time_array([start_date]),
    variable=prognostic.input_coords()["variable"],
    lead_time=prognostic.input_coords()["lead_time"],
    device=device,
)

_ = fetch_data(
    source=prognostic.forcing_data_source,
    time=coords["time"],
    variable=prognostic._forcing_vars_e2s,
    lead_time=np.array([np.timedelta64(0, "h")]),
    device=device,
)

if use_torch_compile:
    if device.type == "cuda":
        torch.cuda.synchronize()
    start_compile = time.perf_counter()
    _ = prognostic(x, coords)
    if device.type == "cuda":
        torch.cuda.synchronize()
    print(f"Torch compile warmup: {time.perf_counter() - start_compile:.6f} seconds")

model = prognostic.create_iterator(x, coords)


nsteps = 28

start_loop = time.perf_counter()

with tqdm(total=nsteps + 1, desc="Running inference", leave=False) as pbar:
    for step, (x_mod, coords_mod) in enumerate(model):
        pbar.update(1)
        if step == nsteps:
            break
end_loop = time.perf_counter()
print(f"Elapsed time(loop[{step}]): {end_loop - start_loop:.6f} seconds")
```

</details>

Validation:
- `uv run pytest test/data/test_ace2_era5.py test/models/px/test_ace2.py`
- `uv run ruff check earth2studio/data/ace2.py earth2studio/models/px/ace2.py test/data/test_ace2_era5.py`

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/earth2studio/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/earth2studio/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/earth2studio/issues) is linked to this pull request.
- [ ] Assess and address Greptile feedback (AI code review bot for guidance; use discretion, addressing all feedback is not required).

## Dependencies

No new dependencies.

### CODE REVIEW

File: earth2studio/models/px/ace2.py [hunk 1]
```diff
@@ -475,18 +538,11 @@ def _forward(
         lead_times = np.array(
             [coords["lead_time"][0], coords["lead_time"][0] + self._dt]
         )
-        forcing_x, forcing_coords = fetch_data(
```
Comment (NickGeneva) [inline earth2studio/models/px/ace2.py:478]: this would re-open a netcdf file every forward

**Comment by NickGeneva** [comment]
@greptile-apps

**Comment by NickGeneva** [comment]
I check consistency with which gave me zero error compared to main.

<details>

```python
#!/usr/bin/env python3
"""Write or compare an ACE2ERA5 multi-step forward reference.

Usage:
  # On main:
  uv run python ace2_forward_consistency.py --write /tmp/ace2-main-forward.npz

  # On the branch under test:
  uv run python ace2_forward_consistency.py --compare /tmp/ace2-main-forward.npz

The artifact is a pickle-free npz containing the output tensor and a JSON-encoded
coordinate summary. By default the script disables TF32 so branch-to-branch
comparisons are not affected by performance precision settings.
"""

from __future__ import annotations

import argparse
import json
from collections.abc import Mapping
from pathlib import Path
from typing import Any

import numpy as np
import torch

from earth2studio.data import fetch_data
from earth2studio.data.ace2 import ACE2ERA5Data
from earth2studio.models.px.ace2 import ACE2ERA5
from earth2studio.utils.time import to_time_array


def configure_torch(*, tf32: bool) -> None:
    torch.backends.cudnn.benchmark = False
    if tf32:
        torch.set_float32_matmul_precision("high")
        torch.backends.cuda.matmul.allow_tf32 = True
        torch.backends.cudnn.allow_tf32 = True
    else:
        torch.set_float32_matmul_precision("highest")
        torch.backends.cuda.matmul.allow_tf32 = False
        torch.backends.cudnn.allow_tf32 = False


def coord_value(value: Any) -> dict[str, Any]:
    arr = np.asarray(value)
    if np.issubdtype(arr.dtype, np.datetime64):
        return {
            "dtype": "datetime64[ns]",
            "values": arr.astype("datetime64[ns]").astype(np.int64).tolist(),
        }
    if np.issubdtype(arr.dtype, np.timedelta64):
        return {
            "dtype": "timedelta64[ns]",
            "values": arr.astype("timedelta64[ns]").astype(np.int64).tolist(),
        }
    if arr.dtype == object or np.issubdtype(arr.dtype, np.str_):
        return {"dtype": "str", "values": arr.astype(str).tolist()}
    return {"dtype": str(arr.dtype), "values": arr.tolist()}


def coords_summary(coords: Mapping[str, Any]) -> dict[str, Any]:
    return {key: coord_value(value) for key, value in coords.items()}


def run_forward(
    start_date: str, device: torch.device, steps: int
) -> tuple[np.ndarray, dict[str, Any]]:
    data = ACE2ERA5Data(mode="initial_conditions", verbose=False)
    package = ACE2ERA5.load_default_package()
    model = ACE2ERA5.load_model(package).to(device)
    model.forcing_data_source._verbose = False

    x, coords = fetch_data(
        source=data,
        time=to_time_array([start_date]),
        variable=model.input_coords()["variable"],
        lead_time=model.input_coords()["lead_time"],
        device=device,
    )

    if steps < 1:
        raise ValueError("steps must be at least 1")

    with torch.inference_mode():
        iterator = model.create_iterator(x, coords)
        next(iterator)  # Consume the initial 0h state; compare forecast steps only.
        outputs = []
        step_coords = []
        for _ in range(steps):
            y_step, out_coords = next(iterator)
            outputs.append(y_step.detach())
            step_coords.append(out_coords.copy())
    if device.type == "cuda":
        torch.cuda.synchronize(device)

    lead_time_axis = list(step_coords[-1]).index("lead_time")
    y = torch.cat(outputs, dim=lead_time_axis)
    out_coords = step_coords[-1].copy()
    out_coords["lead_time"] = np.concatenate(
        [np.asarray(coords["lead_time"]) for coords in step_coords]
    )
    return y.cpu().numpy(), coords_summary(out_coords)


def write_reference(path: Path, y: np.ndarray, coords: dict[str, Any], metadata: dict[str, Any]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    np.savez_compressed(
        path,
        y=y,
        coords=np.array(json.dumps(coords, sort_keys=True)),
        metadata=np.array(json.dumps(metadata, sort_keys=True)),
    )
    print(f"wrote reference: {path}")
    print(f"shape={y.shape} dtype={y.dtype}")


def align_variable_axis(
    y: np.ndarray,
    coords: dict[str, Any],
    reference: dict[str, Any],
) -> tuple[np.ndarray, dict[str, Any]]:
    cur_vars = list(coords.get("variable", {}).get("values", []))
    ref_vars = list(reference.get("variable", {}).get("values", []))
    if cur_vars == ref_vars:
        return y, coords
    if sorted(cur_vars) != sorted(ref_vars):
        return y, coords

    # ACE2 forward output is [time, lead_time, variable, lat, lon] when called
    # without an explicit batch dimension. Align by variable name so hash-order
    # differences between Python processes do not look like numerical changes.
    variable_axis = 2
    indices = [cur_vars.index(name) for name in ref_vars]
    aligned = np.take(y, indices, axis=variable_axis)
    coords = json.loads(json.dumps(coords))
    coords["variable"]["values"] = ref_vars
    return aligned, coords


def compare_coords(
    current: dict[str, Any],
    reference: dict[str, Any],
    *,
    coord_atol: float = 1e-12,
) -> list[str]:
    errors = []
    if current.keys() != reference.keys():
        return [
            "coordinate keys differ: "
            f"current={sorted(current.keys())} reference={sorted(reference.keys())}"
        ]
    for key in current:
        cur = current[key]
        ref = reference[key]
        if cur["dtype"] != ref["dtype"]:
            errors.append(
                f"{key}: dtype differs current={cur['dtype']} reference={ref['dtype']}"
            )
            continue
        if cur["dtype"] in {"datetime64[ns]", "timedelta64[ns]", "str"}:
            if cur["values"] != ref["values"]:
                errors.append(f"{key}: values differ")
            continue
        cur_values = np.asarray(cur["values"], dtype=np.float64)
        ref_values = np.asarray(ref["values"], dtype=np.float64)
        if cur_values.shape != ref_values.shape:
            errors.append(
                f"{key}: shape differs current={cur_values.shape} "
                f"reference={ref_values.shape}"
            )
        elif not np.allclose(cur_values, ref_values, rtol=0.0, atol=coord_atol):
            errors.append(
                f"{key}: numeric values differ "
                f"max_abs={float(np.abs(cur_values - ref_values).max()):.9g}"
            )
    return errors


def compare_reference(
    path: Path,
    y: np.ndarray,
    coords: dict[str, Any],
    *,
    rtol: float,
    atol: float,
) -> None:
    ref = np.load(path, allow_pickle=False)
    y_ref = ref["y"]
    coords_ref = json.loads(str(ref["coords"]))
    y, coords = align_variable_axis(y, coords, coords_ref)

    if y.shape != y_ref.shape:
        raise SystemExit(f"shape mismatch: current={y.shape} reference={y_ref.shape}")
    coord_errors = compare_coords(coords, coords_ref)
    if coord_errors:
        print("coordinate mismatch detected")
        for error in coord_errors:
            print(f"  - {error}")
        raise SystemExit(1)

    diff = np.abs(y.astype(np.float64) - y_ref.astype(np.float64))
    max_abs = float(np.nanmax(diff))
    mean_abs = float(np.nanmean(diff))
    max_ref = float(np.nanmax(np.abs(y_ref.astype(np.float64))))
    passed = bool(np.allclose(y, y_ref, rtol=rtol, atol=atol, equal_nan=True))

    print(f"shape={y.shape} dtype={y.dtype}")
    print(f"max_abs={max_abs:.9g}")
    print(f"mean_abs={mean_abs:.9g}")
    print(f"max_abs_ref={max_ref:.9g}")
    print(f"rtol={rtol} atol={atol}")
    print(f"allclose={passed}")
    if not passed:
        raise SystemExit(1)


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    group = parser.add_mutually_exclusive_group(required=True)
    group.add_argument("--write", type=Path, help="Write reference npz artifact")
    group.add_argument("--compare", type=Path, help="Compare against reference npz artifact")
    parser.add_argument("--start-date", default="1950-01-01")
    parser.add_argument("--steps", type=int, default=5, help="Forecast steps to compare")
    parser.add_argument("--device", default="cuda:0" if torch.cuda.is_available() else "cpu")
    parser.add_argument("--rtol", type=float, default=1e-5)
    parser.add_argument("--atol", type=float, default=1e-5)
    parser.add_argument("--tf32", action="store_true", help="Enable TF32 for the comparison run")
    return parser.parse_args()


def main() -> None:
    args = parse_args()
    device = torch.device(args.device)
    configure_torch(tf32=args.tf32)
    y, coords = run_forward(args.start_date, device, args.steps)
    metadata = {
        "start_date": args.start_date,
        "device": str(device),
        "steps": args.steps,
        "torch_version": torch.__version__,
        "tf32": args.tf32,
        "matmul_precision": torch.get_float32_matmul_precision(),
        "cudnn_allow_tf32": torch.backends.cudnn.allow_tf32,
        "cuda_matmul_allow_tf32": torch.backends.cuda.matmul.allow_tf32,
    }
    print(json.dumps(metadata, indent=2, sort_keys=True))
    if args.write is not None:
        write_reference(args.write, y, coords, metadata)
    else:
        compare_reference(args.compare, y, coords, rtol=args.rtol, atol=args.atol)


if __name__ == "__main__":
    main()

```

</details>

File: earth2studio/models/px/ace2.py [hunk 2]
```diff
@@ -446,6 +462,123 @@ def _batch_data_to_tensor(self, data: dict[str, torch.Tensor]) -> torch.Tensor:
         y = y.unsqueeze(2)
         return y
 
+    def _fetch_forcing_year(
```
Comment (NickGeneva) [inline earth2studio/models/px/ace2.py:470]: the forcing data is actually pretty small, so we will just cache a full year it on the gpu

0.5 Gb for the whole year

File: earth2studio/models/px/ace2.py [hunk 3]
```diff
@@ -446,6 +462,112 @@ def _batch_data_to_tensor(self, data: dict[str, torch.Tensor]) -> torch.Tensor:
...
+        self, valid_time: np.datetime64, device: torch.device
+    ) -> tuple[torch.Tensor, CoordSystem]:
+        valid_time = valid_time.astype("datetime64[ns]")
+        device = torch.device(device)
+        if isinstance(self.forcing_data_source, ACE2ERA5Data):
+            year = pd.Timestamp(valid_time).year
+            forcing_x, forcing_coords = self._fetch_forcing_year(year, device)
+
+            year_start = np.datetime64(f"{year:04d}-01-01T00:00:00", "ns")
+            delta_ns = (
```
Comment (NickGeneva) [inline earth2studio/models/px/ace2.py:516]: calc the year offset in ns, then compute steps (index in the year netcdf) for the slice we need to read

File: earth2studio/models/px/ace2.py [hunk 4]
```diff
@@ -446,6 +462,112 @@ def _batch_data_to_tensor(self, data: dict[str, torch.Tensor]) -> torch.Tensor:
...
+
+        cache_key = (
+            id(self.forcing_data_source),
+            tuple(self._forcing_vars_e2s),
+            str(device),
+            "time",
+            int(valid_time.astype("datetime64[ns]").astype(np.int64)),
+        )
+        if cache_key not in self._forcing_cache:
+            self._forcing_cache[cache_key] = fetch_data(
```
Comment (NickGeneva) [inline earth2studio/models/px/ace2.py:538]: the fallback for other data sources is not the ace2 one

File: test/data/test_ace2_era5.py [hunk 5]
```diff
@@ -139,6 +141,48 @@ def test_ace2era5_initial_conditions_year_validation():
         ds(datetime.datetime(year=1981, month=1, day=1, hour=0), ["skt"])
 
 
+def test_ace2era5_forcing_static_fields_expand_multiple_times(monkeypatch, tmp_path):
```
Comment (NickGeneva) [inline test/data/test_ace2_era5.py:144]: same additional test provided by codex

File: earth2studio/models/px/ace2.py [hunk 6]
```diff
@@ -128,8 +129,19 @@ class ACE2ERA5(torch.nn.Module, AutoModelMixin, PrognosticMixin):
...
     - ACE2 code: https://github.com/ai2cm/ace
+    - Huggingface: https://huggingface.co/allenai/ACE2-ERA5
+
+    Notes
+    -----
+    For throughput-sensitive GPU inference, enabling TensorFloat-32 matmul kernels
+    before running ACE2 can improve performance on supported NVIDIA GPUs:
+
+        torch.set_float32_matmul_precision("high")
+
```
Comment (pzharrington) [inline earth2studio/models/px/ace2.py:147]: This is the same setting as what we do for Atlas, right? If so we might as well be consistent, I think that one we say to set performance based on env variable rather than explicit Python call

**Review by pzharrington — APPROVED** [review]

File: earth2studio/models/px/ace2.py (hunk 6 above)
Comment (NickGeneva) [reply earth2studio/models/px/ace2.py:147]: oh is it? Okay let me update this

Comment (NickGeneva) [reply earth2studio/models/px/ace2.py:147]: added comment in api doc. TORCH_ALLOW_TF32_CUBLAS_OVERRIDE=1 is more force full than the torch setting. but also gets the job done
